distributed/worker_state_machine.py

```diff
@@ -1970,214 +1944,183 @@ def _transition_executing_error(
+            ts.previous = None
+        elif self.validate:
+            assert ts.previous == "flight"
+            assert ts.next == "waiting"
+
+        return {}, []
 
     def _transition_cancelled_fetch(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        if ts.done:
-            return {ts: "released"}, []
-        elif ts.previous == "flight":
-            ts.state = ts.previous
-            return {}, []
+        """
+        See also
+        --------
+        _transition_cancelled_waiting
+        _transition_resumed_fetch
+        _transition_resumed_waiting
+        """
+        if ts.previous == "flight":
+            if ts.done:
+                # gather_dep just completed for a cancelled task.
+                # Discard output and possibly forget
+                return {ts: "released"}, []
+            else:
+                # Forget the task was cancelled to begin with
+                ts.state = "flight"
+                ts.previous = None
+                return {}, []
         else:
             assert ts.previous in ("executing", "long-running")
+            # None of the exit events of execute recommend a transition to fetch
+            assert not ts.done
             ts.state = "resumed"
             ts.next = "fetch"
             return {}, []
 
     def _transition_cancelled_waiting(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        if ts.done:
-            return {ts: "released"}, []
-        elif ts.previous in ("executing", "long-running"):
+        """
+        See also
+        --------
+        _transition_cancelled_fetch
+        _transition_resumed_fetch
+        _transition_resumed_waiting
+        """
+        # None of the exit events of gather_dep or execute recommend a transition to
+        # waiting
+        assert not ts.done
+        if ts.previous in ("executing", "long-running"):
+            # Forget the task was cancelled to begin with
             ts.state = ts.previous
+            ts.previous = None
             return {}, []
         else:
             assert ts.previous == "flight"
             ts.state = "resumed"
             ts.next = "waiting"
             return {}, []
 
-    def _transition_cancelled_forgotten(
-        self, ts: TaskState, *, stimulus_id: str
-    ) -> RecsInstrs:
-        ts.next = "forgotten"
-        if not ts.done:
-            return {}, []
-        return {ts: "released"}, []
-
     def _transition_cancelled_released(
-        self, ts: TaskState, *, stimulus_id: str
+        self,
+        ts: TaskState,
+        *args: Any,  # extra arguments of transitions to memory or error - ignored
+        stimulus_id: str,
     ) -> RecsInstrs:
         if not ts.done:
             return {}, []
-        self.executing.discard(ts)
-        self.long_running.discard(ts)
-        self.in_flight_tasks.discard(ts)
 
-        self._release_resources(ts)
+        ts.previous = None
+        ts.done = False
```

Comment:
This is not necessary. The generic_release triggers a purge_state which rests this

```suggestion
```

Reply:
Yes, this was a conscious decision. I'd like to be as explicit as possible in the specific code.